### ISSUE TITLE
contract verification ordering degraded

### ISSUE THREAD

**Issue by wminshew** [foundry-rs/foundry#3553]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (9d64f1f 2022-10-26T00:11:13.985133Z)

### What command(s) is the bug in?

forge verify-contract

### Operating System

macOS (Intel)

### Describe the bug

it looks like somewhat recently (last ~20 days?) the ordering for verified contracts changed from contract / import order to.. something that's almost alphabetical but somehow even worse?

old/good example @ https://etherscan.io/address/0x4Df01754eBd055498C8087b1e9a5c7a9ad19b0F6#code
new/bad example @ https://arbiscan.io/address/0x4Df01754eBd055498C8087b1e9a5c7a9ad19b0F6#code

the former is far more useful/readable imo

**Comment by mattsse** [comment]
cc @rkrasiuk anything comes to mind that might have caused this?

**Comment by rkrasiuk** [comment]
@mattsse nothing obvious comes to mind. dyk of any graph resolution changes in ethers that might've caused this? also, arbiscan is hella weird by itself, not a 1-to-1 port of etherscan. @wminshew any chance you observed this on other explorers?

**Comment by wminshew** [comment]
@rkrasiuk yes, can see everything deployed/verified by foundry in a more recent repo mirrors the new/bad arbiscan pattern above:

[repo](https://github.com/0xSplits/splits-liquid)
[etherscan](https://etherscan.io/address/0xdecd8b99b7f763e16141450daa5ea414b7994831#code)
[other flavors of etherscan](https://blockscan.com/address/0xdecd8b99b7f763e16141450daa5ea414b7994831)

potentially related / worth mentioning, in both repos (waterfall in the og links & liquid in the above links) the `--verify` flag wasn't working w `forge script` (I assume bc of deterministic deployment? but I didn't dig into it) so all of these were verified w separate `verify-contract` calls

**Comment by wminshew** [comment]
appears to still be the case https://gnosisscan.io/address/0x1a1314121b06f1e358eec90e76e459a679c88863#code

(also `--verify` still doesn't work w script)

**Comment by mpeyfuss** [comment]
I looked into this a bit more since it's been bugging me. Not super familiar with rust so I can't pinpoint exactly what's going on but this is what I've found.

I had a friend using commit `1d9a34e` deploy and verify a multipart contract and it resulted in a good example. Same commit on my machine resulted in a bad example for the same contract.

Looking at the build info (JSON standard input), it looks like the sources are ordered opposite. The build info was retrieved using `forge build --build-info`.

It feels crazy but is there machine dependent ordering of the source list in the JSON standard input? Block explorers take the list supplied as is and that determines the order of the files shown.

I'll keep looking into this but hopefully this sparks some inspiration and hopefully we can close out this issue. Perhaps @rkrasiuk this helps 🤞
